### ISSUE TITLE
Only on Windows, the default gem is preferred over the installed gem

### ISSUE THREAD

**Issue by nobu** [rubygems/rubygems#7726]
### Describe the problem as clearly as you can

After installed a built gem that is the same version with the default gem, 
the default gem is loaded, not the installed gem.

This happens only on Windows, not Ubuntu/macOS.

#### In ruby/strscan CI

[After installed the built gem](https://github.com/ruby/strscan/actions/runs/9362666795/job/25771858572#step:10:1):
```
current directory: D:/ruby-mingw/lib/ruby/gems/3.4.0+0/gems/strscan-3.1.2/ext/strscan
D:/ruby-mingw/bin/ruby.exe extconf.rb
current directory: D:/ruby-mingw/lib/ruby/gems/3.4.0+0/gems/strscan-3.1.2/ext/strscan
make.exe DESTDIR\= sitearchdir\=./.gem.20240604-560-dlqm6z sitelibdir\=./.gem.20240604-560-dlqm6z clean
current directory: D:/ruby-mingw/lib/ruby/gems/3.4.0+0/gems/strscan-3.1.2/ext/strscan
make.exe DESTDIR\= sitearchdir\=./.gem.20240604-560-dlqm6z sitelibdir\=./.gem.20240604-560-dlqm6z
current directory: D:/ruby-mingw/lib/ruby/gems/3.4.0+0/gems/strscan-3.1.2/ext/strscan
make.exe DESTDIR\= sitearchdir\=./.gem.20240604-560-dlqm6z sitelibdir\=./.gem.20240604-560-dlqm6z install
current directory: D:/ruby-mingw/lib/ruby/gems/3.4.0+0/gems/strscan-3.1.2/ext/strscan
make.exe DESTDIR\= sitearchdir\=./.gem.20240604-560-dlqm6z sitelibdir\=./.gem.20240604-560-dlqm6z clean
Successfully installed strscan-3.1.2
1 gem installed
```

[The default gem is loaded but not the installed gem](https://github.com/ruby/strscan/actions/runs/9362666795/job/25771858572#step:12:24).

```
https://github.com/ruby/strscan/actions/runs/9362666795/job/25771858572#step:12:24
Loaded strscan from D:/ruby-mingw/lib/ruby/gems/3.4.0+0/gems/strscan-3.1.2/lib/strscan.so
Gem from D:/ruby-mingw/lib/ruby/gems/3.4.0+0/gems/strscan-3.1.2
```

It works as expected on [Ubuntu] and [macOS].

[Ubuntu]: https://github.com/ruby/strscan/actions/runs/9362666795/job/25771855889#step:12:5
[macOS]: https://github.com/ruby/strscan/actions/runs/9362666795/job/25771856097#step:12:7

#### In ruby/digest CI

(Similar)[https://github.com/nobu/digest/actions/runs/9433378760/job/25984253681#step:9:24] to strscan.
```
gem install --install-dir D:/a/_temp/d20240609-5128-3w4he9 --no-document pkg/digest-3.1.1.gem
Temporarily enhancing PATH for MSYS/MINGW...
Using msys2 packages: mingw-w64-ucrt-x86_64-openssl
Building native extensions. This could take a while...
Successfully installed digest-3.1.1
1 gem installed
Unexpected digest is loaded: ["D:/ruby-ucrt/lib/ruby/3.4.0+0/x64-mingw-ucrt/digest.so"]
digest 3.1.1 is loaded.
check failed!
```

This ruby is invoked with `{"GEM_HOME"=>tmpdir}` environment variable, where `tmpdir` is for the `--install-dir` option.

### Did you try upgrading RubyGems?

Yes, on [ruby/digest](https://github.com/nobu/digest/actions/runs/9433378760/job/25984253681#step:3:57).

### Post steps to reproduce the problem

* Installed ruby using ruby/setup-ruby action.
* RubyGems is configured by the action.
* As the above, only on Windows, the latest Ruby and a gem that is the same version with a default gem.

#### In ruby/strscan CI

```
gem install --verbose --backtrace pkg/*.gem
ruby run-test.rb
```

The part corresponding to the output in `run-test.rb` is:
```ruby
gem 'strscan'
require 'strscan'
puts "Loaded strscan from #{$".grep(/\/strscan\./).join(', ')}"
puts "Gem from #{Gem.loaded_specs["strscan"]&.full_gem_path}"
```

#### In ruby/digest CI
```
bundle exec rake check
```

And the `check` task is:

```ruby
      tmpdir = File.realpath(tmpdir)
      sh "gem", "install", "--install-dir", tmpdir, "--no-document", gem

      _, _, status = EnvUtil.invoke_ruby([{"GEM_HOME"=>tmpdir}], <<~EOS)
      version = #{version.dump}
      gem "digest", version
      loaded_version = Gem.loaded_specs["digest"].version.to_s

      if loaded_version == version
        puts "digest \#{loaded_version} is loaded."
      else
        abort "digest \#{loaded_version} is loaded instead of \#{version}!"
      end
```

### What were you expecting to happen?

The installed gems are used on Windows as well as on Ubuntu/macOS.

### What actually happened?

The default gem is loaded, not the installed gem.

### Run `gem env` and paste the output below

[Output in ruby/digest](https://github.com/nobu/digest/actions/runs/9433378760/job/25984253681#step:8:1):

```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.6.0.dev
  - RUBY VERSION: 3.4.0 (2024-06-09 patchlevel -1) [x64-mingw-ucrt]
  - INSTALLATION DIRECTORY: D:/ruby-ucrt/lib/ruby/gems/3.4.0+0
  - USER INSTALLATION DIRECTORY: C:/Users/runneradmin/.local/share/gem/ruby/3.4.0+0
  - RUBY EXECUTABLE: D:/ruby-ucrt/bin/ruby.exe
  - GIT EXECUTABLE: C:\Program Files\Git\bin/git.EXE
  - EXECUTABLE DIRECTORY: D:/ruby-ucrt/bin
  - SPEC CACHE DIRECTORY: C:/Users/runneradmin/.cache/gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: C:/ProgramData
  - RUBYGEMS PLATFORMS:
     - ruby
     - x64-mingw-ucrt
  - GEM PATHS:
     - D:/ruby-ucrt/lib/ruby/gems/3.4.0+0
     - C:/Users/runneradmin/.local/share/gem/ruby/3.4.0+0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => true
     - :bulk_threshold => 1000
     - "# Feel free to add any rubygems config options as described on" => nil
     - "gem" => "--no-document"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - C:\Program Files\PowerShell\7
     - D:\ruby-ucrt\bin
     - C:\msys64\ucrt64\bin
     - C:\msys64\usr\bin
     - C:\Program Files\MongoDB\Server\5.0\bin
     - C:\aliyun-cli
     - C:\vcpkg
     - C:\Program Files (x86)\NSIS\
     - C:\tools\zstd
     - C:\Program Files\Mercurial\
     - C:\hostedtoolcache\windows\stack\2.15.7\x64
     - C:\cabal\bin
     - C:\\ghcup\bin
     - C:\mingw64\bin
     - C:\Program Files\dotnet
     - C:\Program Files\MySQL\MySQL Server 8.0\bin
     - C:\Program Files\R\R-4.4.0\bin\x64
     - C:\SeleniumWebDrivers\GeckoDriver
     - C:\SeleniumWebDrivers\EdgeDriver\
     - C:\SeleniumWebDrivers\ChromeDriver
     - C:\Program Files (x86)\sbt\bin
     - C:\Program Files (x86)\GitHub CLI
     - C:\Program Files\Git\bin
     - C:\Program Files (x86)\pipx_bin
     - C:\npm\prefix
     - C:\hostedtoolcache\windows\go\1.21.10\x64\bin
     - C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts
     - C:\hostedtoolcache\windows\Python\3.9.13\x64
     - C:\Program Files\OpenSSL\bin
     - C:\tools\kotlinc\bin
     - C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.412-8\x64\bin
     - C:\Program Files\ImageMagick-7.1.1-Q16-HDRI
     - C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin
     - C:\ProgramData\kind
     - C:\ProgramData\docker-compose
     - C:\ProgramData\Chocolatey\bin
     - C:\Windows\system32
     - C:\Windows
     - C:\Windows\System32\Wbem
     - C:\Windows\System32\WindowsPowerShell\v1.0\
     - C:\Windows\System32\OpenSSH\
     - C:\Program Files\dotnet\
     - C:\Program Files\PowerShell\7\
     - C:\Program Files\Microsoft\Web Platform Installer\
     - C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\
     - C:\Program Files\Microsoft SQL Server\150\Tools\Binn\
     - C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\
     - C:\Program Files (x86)\WiX Toolset v3.14\bin
     - C:\Program Files\Microsoft SQL Server\130\DTS\Binn\
     - C:\Program Files\Microsoft SQL Server\140\DTS\Binn\
     - C:\Program Files\Microsoft SQL Server\150\DTS\Binn\
     - C:\Program Files\Microsoft SQL Server\160\DTS\Binn\
     - C:\Strawberry\c\bin
     - C:\Strawberry\perl\site\bin
     - C:\Strawberry\perl\bin
     - C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin
     - C:\Program Files\TortoiseSVN\bin
     - C:\Program Files\CMake\bin
     - C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.7\bin
     - C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code
     - C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager
     - C:\Program Files\nodejs\
     - C:\Program Files\Git\cmd
     - C:\Program Files\Git\mingw64\bin
     - C:\Program Files\Git\usr\bin
     - C:\Program Files\GitHub CLI\
     - c:\tools\php
     - C:\Program Files (x86)\sbt\bin
     - C:\Program Files\Amazon\AWSCLIV2\
     - C:\Program Files\Amazon\SessionManagerPlugin\bin\
     - C:\Program Files\Amazon\AWSSAMCLI\bin\
     - C:\Program Files\Microsoft SQL Server\130\Tools\Binn\
     - C:\Program Files\LLVM\bin
     - C:\Users\runneradmin\.dotnet\tools
     - C:\Users\runneradmin\.cargo\bin
     - C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps
```

**Comment by MSP-Greg** [comment]
I tried this locally, and everything was ok.  But, my `.gemrc` file uses 'user-install'. I then installed the gem in the standard location, and I could repro the issue.

https://github.com/rubygems/rubygems/blob/11a1ac52b550cf485d1a56e4946c9d619892d347/lib/rubygems/specification.rb#L835

I changed to:
```ruby
b.base_dir == Gem.path.first || !b.default_gem? ? 1 : -1
```
and it works correctly.  With the gem installed in all three-locations (user, standard, & default), user is activated.

@deivid-rodriguez This is an update to #7700, which both of us were involved with...
